File: apps/web/src/utils/fabraryCsv.js
/**
 * RFC4180 reader for a Fabrary collection export.
 * Required headers are case-sensitive as exported.
 */

export const FABRARY_REQUIRED_HEADERS = [
    'Identifier',
    'Name',
    'Pitch',
    'Set',
    'Set number',
    'Edition',
    'Foiling',
    'Treatment',
    'Have',
];

/**
 * Split a CSV string into a header row + records. Quoted fields may contain
 * commas and escaped quotes (`""`).
 *
 * @param {string} text
 * @returns {string[][]}
 */
export function parseCsvRecords(text) {
    const input = String(text ?? '').replace(/^\uFEFF/, '');
    const rows = [];
    let row = [];
    let field = '';
    let i = 0;
    let inQuotes = false;

    const pushField = () => {
        row.push(field);
        field = '';
    };
    const pushRow = () => {
        // A trailing newline should not invent an empty record.
        if (row.length === 1 && row[0] === '' && !inQuotes) {
            row = [];
            return;
        }
        rows.push(row);
        row = [];
    };

    while (i < input.length) {
        const ch = input[i];
        if (inQuotes) {
            if (ch === '"') {
                if (input[i + 1] === '"') {
                    field += '"';
                    i += 2;
                    continue;
                }
                inQuotes = false;
                i += 1;
                continue;
            }
            field += ch;
            i += 1;
            continue;
        }
        if (ch === '"') {
            inQuotes = true;
            i += 1;
            continue;
        }
        if (ch === ',') {
            pushField();
            i += 1;
            continue;
        }
        if (ch === '\n') {
            pushField();
            pushRow();
            i += 1;
            continue;
        }
        if (ch === '\r') {
            i += 1;
            continue;
        }
        field += ch;
        i += 1;
    }
    if (inQuotes || field !== '' || row.length > 0) {
        pushField();
        pushRow();
    }
    return rows;
}

/**
 * @param {string} text
 * @returns {{ ok: boolean, reason?: string, headers: string[], rows: Object[] }}
 */
export function parseFabraryCsv(text) {
    const records = parseCsvRecords(text);
    if (records.length === 0) {
        return { ok: false, reason: 'not_fabrary', headers: [], rows: [] };
    }
    const headers = records[0].map((h) => String(h ?? ''));
    const missing = FABRARY_REQUIRED_HEADERS.filter((h) => !headers.includes(h));
    if (missing.length > 0) {
        return { ok: false, reason: 'not_fabrary', headers, rows: [] };
    }
    const rows = records.slice(1).map((cells) => {
        const row = {};
        headers.forEach((header, index) => {
            row[header] = cells[index] ?? '';
        });
        return row;
    });
    return { ok: true, headers, rows };
}

export function hasFabraryHeaders(headers) {
    const list = headers || [];
    return FABRARY_REQUIRED_HEADERS.every((h) => list.includes(h));
}

File: apps/mobile/lib/core/logic/fabrary_csv.dart
/// RFC4180 reader for a Fabrary collection export.
const fabraryRequiredHeaders = [
  'Identifier',
  'Name',
  'Pitch',
  'Set',
  'Set number',
  'Edition',
  'Foiling',
  'Treatment',
  'Have',
];

class FabraryCsvParse {
  const FabraryCsvParse({
    required this.ok,
    this.reason,
    required this.headers,
    required this.rows,
  });

  final bool ok;
  final String? reason;
  final List<String> headers;
  final List<Map<String, String>> rows;
}

/// Split a CSV string into records. Quoted fields may contain commas.
List<List<String>> parseCsvRecords(String text) {
  final input = text.startsWith('\uFEFF') ? text.substring(1) : text;
  final rows = <List<String>>[];
  var row = <String>[];
  var field = StringBuffer();
  var i = 0;
  var inQuotes = false;

  void pushField() {
    row.add(field.toString());
    field = StringBuffer();
  }

  void pushRow() {
    if (row.length == 1 && row[0].isEmpty && !inQuotes) {
      row = <String>[];
      return;
    }
    rows.add(row);
    row = <String>[];
  }

  while (i < input.length) {
    final ch = input[i];
    if (inQuotes) {
      if (ch == '"') {
        if (i + 1 < input.length && input[i + 1] == '"') {
          field.write('"');
          i += 2;
          continue;
        }
        inQuotes = false;
        i += 1;
        continue;
      }
      field.write(ch);
      i += 1;
      continue;
    }
    if (ch == '"') {
      inQuotes = true;
      i += 1;
      continue;
    }
    if (ch == ',') {
      pushField();
      i += 1;
      continue;
    }
    if (ch == '\n') {
      pushField();
      pushRow();
      i += 1;
      continue;
    }
    if (ch == '\r') {
      i += 1;
      continue;
    }
    field.write(ch);
    i += 1;
  }
  if (inQuotes || field.isNotEmpty || row.isNotEmpty) {
    pushField();
    pushRow();
  }
  return rows;
}

bool hasFabraryHeaders(Iterable<String>? headers) {
  final list = headers?.toList() ?? const <String>[];
  return fabraryRequiredHeaders.every(list.contains);
}

FabraryCsvParse parseFabraryCsv(String text) {
  final records = parseCsvRecords(text);
  if (records.isEmpty) {
    return const FabraryCsvParse(
      ok: false,
      reason: 'not_fabrary',
      headers: [],
      rows: [],
    );
  }
  final headers = records.first.map((h) => h).toList();
  final missing =
      fabraryRequiredHeaders.where((h) => !headers.contains(h)).toList();
  if (missing.isNotEmpty) {
    return FabraryCsvParse(
      ok: false,
      reason: 'not_fabrary',
      headers: headers,
      rows: const [],
    );
  }
  final rows = records.skip(1).map((cells) {
    final row = <String, String>{};
    for (var i = 0; i < headers.length; i++) {
      row[headers[i]] = i < cells.length ? cells[i] : '';
    }
    return row;
  }).toList();
  return FabraryCsvParse(ok: true, headers: headers, rows: rows);
}
